tx-compiler/src/compiler.rs

```diff
@@ -0,0 +1,299 @@
+    Program, ProgramAst, StdLibrary, TransactionError, TransactionKernel,
+};
+
+// TRANSACTION COMPILER
+// ================================================================================================
+
+pub struct TransactionComplier {
+    assembler: Assembler,
+    account_procedures: HashMap<AccountId, Vec<Digest>>,
+    prologue: CodeBlock,
+    epilogue: CodeBlock,
+    note_setup: CodeBlock,
+    tx_kernel: Kernel,
+}
+
+impl TransactionComplier {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    pub fn new() -> TransactionComplier {
+        let (tx_kernel, tx_module) = TransactionKernel::kernel();
+        let assembler = Assembler::default()
+            .with_library(&MidenLib::default())
+            .expect("library is well formed")
+            .with_library(&StdLibrary::default())
+            .expect("library is well formed")
+            .with_kernel_module(tx_module)
+            .expect("kernel is well formed");
+        TransactionComplier {
+            assembler,
+            account_procedures: HashMap::default(),
+            prologue: TransactionKernel::prologue(),
+            epilogue: TransactionKernel::epilogue(),
+            note_setup: TransactionKernel::note_setup(),
+            tx_kernel,
+        }
+    }
+
+    /// Compiles the provided module into [AccountCode] and associates the resulting procedures
+    /// with the specified account ID.
+    pub fn load_account(
+        &mut self,
+        account_id: AccountId,
+        account_code: ModuleAst,
+    ) -> Result<AccountCode, TransactionError> {
+        let account_code = AccountCode::new(account_code, account_id, &mut self.assembler)
+            .map_err(TransactionError::LoadAccountFailed)?;
+        self.account_procedures.insert(account_id, account_code.procedures().to_vec());
+        Ok(account_code)
+    }
+
+    /// Loads the provided account interface (vector of procedure digests) into the `account_procedures` map.
+    /// Returns the old account interface if it previously existed.
+    pub fn load_account_interface(
+        &mut self,
+        account_id: AccountId,
+        procedures: Vec<Digest>,
+    ) -> Option<Vec<Digest>> {
+        self.account_procedures.insert(account_id, procedures)
+    }
+
+    /// Compiles the provided program into the [NoteScript] and checks (to the extent possible)
+    /// if a note could be executed against an account with the specified interface.
+    pub fn compile_note_script(
+        &mut self,
+        note_script_ast: ProgramAst,
+        // TODO: Should this be optional - what if we don't know the target account interface.
+        target_account_proc: Vec<NoteTarget>,
+    ) -> Result<NoteScript, TransactionError> {
+        let (note_script, code_block) = NoteScript::new(
+            note_script_ast,
+            &mut self.assembler,
+            &mut AssemblyContext::new(AssemblyContextType::Program),
```

Comment:
Is there a reason to pass `AssemblyContext` as an argument here rather creating it in `NoteScript::new()`?

Reply:
No, I have internalised it now.